intellij/org.eclipse.xtend.idea/src/org/eclipse/xtend/core/idea/config/XtendSupportConfigurable.xtend

```diff
@@ -33,18 +35,46 @@ class XtendSupportConfigurable extends FrameworkSupportInModuleConfigurable {
 		ModifiableRootModel rootModel,
 		ModifiableModelsProvider modifiableModelsProvider
 	) {
-		val output = new File(new File(module.moduleFilePath).parentFile, 'xtend-gen')
-		val url = output.path.toSystemIndependentName.pathToUrl
-		rootModel.contentEntries.findFirst[!sourceFolders.empty].addSourceFolder(url, false)
-		if (!output.exists && output.mkdirs) {
-			LocalFileSystem.instance.refreshAndFindFileByIoFile(output)
+		val entry = rootModel.contentEntries.head
+		val mainSrc = entry.sourceFolders.filter[!testSource].head
+		var VirtualFile xtendGenMain = null
+		if (mainSrc != null) {
+			xtendGenMain = mainSrc.file.parent.getOrCreateDir("xtend-gen")
+			val properties = JpsJavaExtensionService.getInstance().createSourceRootProperties("", true);
+			entry.addSourceFolder(xtendGenMain, JavaSourceRootType.SOURCE, properties)
+		}
+		val testSrc = entry.sourceFolders.filter[testSource].head
+		var VirtualFile xtendGenTest = null
+		if (testSrc != null) {
+			xtendGenTest = testSrc.file.parent.getOrCreateDir("xtend-gen")
+			val properties = JpsJavaExtensionService.getInstance().createSourceRootProperties("", true);
+			entry.addSourceFolder(xtendGenTest, JavaSourceRootType.TEST_SOURCE, properties)
 		}
+		
+		val facetType = FacetTypeRegistry.getInstance().findFacetType(XtendLanguage.INSTANCE.ID)
+		val mnr = FacetManager.getInstance(module)
+		var facet = mnr.findFacet(facetType.id, facetType.defaultFacetName) 
+						?: FacetManager.getInstance(module).addFacet(facetType, facetType.defaultFacetName, null)
+		val conf = facet.configuration as XtendFacetConfiguration
+		val state = conf.state
+		if (xtendGenMain != null)
+			state.outputDirectory = xtendGenMain.canonicalPath
+		if (xtendGenTest != null)
+			state.testOutputDirectory = xtendGenTest.canonicalPath
 		val libraryTable = LibraryTablesRegistrar.getInstance().getLibraryTable()
 		var library = libraryTable.getLibraryByName(XtendLibraryDescription.XTEND_LIBRARY_NAME)
 		if (library != null && !rootModel.moduleLibraryTable.libraries.contains(library)) {
 			rootModel.addLibraryEntry(library)
 		}
 	}
+	
+	private def VirtualFile getOrCreateDir(VirtualFile parent, String name) {
+		val existing = parent.children.findFirst[it.name==name] 
```

Comment:
parent.findOrCreateChildData(..) - the requestor should also not be null according to the docs